messaging/messaging.go

```diff
@@ -0,0 +1,454 @@
+	}
+	return c.makeSendRequest(ctx, payload)
+}
+
+// SubscribeToTopic subscribes a list of registration tokens to a topic.
+//
+// The tokens list must not be empty, with at most 1000 tokens.
+func (c *Client) SubscribeToTopic(ctx context.Context, tokens []string, topic string) (*TopicManagementResponse, error) {
+	req := &iidRequest{
+		Topic:  topic,
+		Tokens: tokens,
+		op:     iidSubscribe,
+	}
+	return c.makeTopicManagementRequest(ctx, req)
+}
+
+// UnsubscribeFromTopic unsubscribes a list of registration tokens from a topic.
+//
+// The tokens list must not be empty, with at most 1000 tokens.
+func (c *Client) UnsubscribeFromTopic(ctx context.Context, tokens []string, topic string) (*TopicManagementResponse, error) {
+	req := &iidRequest{
+		Topic:  topic,
+		Tokens: tokens,
+		op:     iidSubscribe,
+	}
+	return c.makeTopicManagementRequest(ctx, req)
+}
+
+type fcmRequest struct {
+	ValidateOnly bool     `json:"validate_only,omitempty"`
+	Message      *Message `json:"message,omitempty"`
+}
+
+type fcmResponse struct {
+	Name string `json:"name"`
+}
+
+type fcmError struct {
+	Error struct {
+		Status string `json:"status"`
+	} `json:"error"`
+}
+
+type iidRequest struct {
+	Topic  string   `json:"to"`
+	Tokens []string `json:"registration_tokens"`
+	op     string
+}
+
+type iidResponse struct {
+	Results []map[string]interface{} `json:"results"`
+}
+
+type iidError struct {
+	Error string `json:"error"`
+}
+
+func (c *Client) makeSendRequest(ctx context.Context, req *fcmRequest) (string, error) {
+	if err := validateMessage(req.Message); err != nil {
+		return "", err
+	}
+
+	request := &internal.Request{
+		Method: http.MethodPost,
+		URL:    fmt.Sprintf("%s/projects/%s/messages:send", c.fcmEndpoint, c.project),
+		Body:   internal.NewJSONEntity(req),
+	}
+	resp, err := c.client.Do(ctx, request)
+	if err != nil {
+		return "", err
+	}
+
+	if resp.Status == http.StatusOK {
+		var result fcmResponse
+		err := json.Unmarshal(resp.Body, &result)
+		return result.Name, err
+	}
+
+	var fe fcmError
+	json.Unmarshal(resp.Body, &fe) // ignore any json parse errors at this level
+	msg := fcmErrorCodes[fe.Error.Status]
+	if msg == "" {
+		msg = fmt.Sprintf("client encounterd an unknown error; response: %s", string(resp.Body))
+	}
+	return "", fmt.Errorf("http error status: %d; reason: %s", resp.Status, msg)
+}
+
+func (c *Client) makeTopicManagementRequest(ctx context.Context, req *iidRequest) (*TopicManagementResponse, error) {
+	if req.Tokens == nil {
+		return nil, fmt.Errorf("no tokens specified")
+	}
+	if len(req.Tokens) > 1000 {
+		return nil, fmt.Errorf("tokens list must not contain more than 1000 items")
+	}
+	for _, token := range req.Tokens {
+		if token == "" {
+			return nil, fmt.Errorf("tokens list must not contain empty strings")
+		}
+	}
+
+	if req.Topic == "" {
+		return nil, fmt.Errorf("topic name not specified")
+	}
+	if !regexp.MustCompile("^(/topics/)?(private/)?[a-zA-Z0-9-_.~%]+$").MatchString(req.Topic) {
```

Comment:
How about pulling this into a var so we don't re-run MustCompile every time this is called?

Reply:
Done, good idea.